golem-worker-executor-base/src/durable_host/rdbms/mysql.rs

```diff
@@ -0,0 +1,1017 @@
+                .as_wasi_view()
+                .table()
+                .get::<MysqlDbConnection>(&self_)?
+                .pool_key
+                .clone();
+            let (input, result) = match to_db_values(params) {
+                Ok(params) => {
+                    let result = self
+                        .state
+                        .rdbms_service
+                        .mysql()
+                        .query(&pool_key, &worker_id, &statement, params.clone())
+                        .await;
+                    (
+                        Some(RdbmsRequest::<MysqlType>::new(pool_key, statement, params)),
+                        result,
+                    )
+                }
+                Err(error) => (None, Err(RdbmsError::QueryParameterFailure(error))),
+            };
+            durability.persist(self, input, result).await
+        } else {
+            durability.replay(self).await
+        };
+        Ok(result.map(DbResult::from).map_err(Error::from))
+    }
+
+    async fn execute(
+        &mut self,
+        self_: Resource<MysqlDbConnection>,
+        statement: String,
+        params: Vec<DbValue>,
+    ) -> anyhow::Result<Result<u64, Error>> {
+        let worker_id = self.state.owned_worker_id.worker_id.clone();
+        let durability = Durability::<u64, SerializableError>::new(
+            self,
+            "rdbms::mysql::db-connection",
+            "execute",
+            DurableFunctionType::WriteRemote,
+        )
+        .await?;
+
+        let result = if durability.is_live() {
+            let pool_key = self
+                .as_wasi_view()
+                .table()
+                .get::<MysqlDbConnection>(&self_)?
+                .pool_key
+                .clone();
+
+            let (input, result) = match to_db_values(params) {
+                Ok(params) => {
+                    let result = self
+                        .state
+                        .rdbms_service
+                        .mysql()
+                        .execute(&pool_key, &worker_id, &statement, params.clone())
+                        .await;
+                    (
+                        Some(RdbmsRequest::<MysqlType>::new(pool_key, statement, params)),
+                        result,
+                    )
+                }
+                Err(error) => (None, Err(RdbmsError::QueryParameterFailure(error))),
+            };
+            durability.persist(self, input, result).await
+        } else {
+            durability.replay(self).await
+        };
+        Ok(result.map_err(Error::from))
+    }
+
+    async fn begin_transaction(
+        &mut self,
+        self_: Resource<MysqlDbConnection>,
+    ) -> anyhow::Result<Result<Resource<DbTransactionEntry>, Error>> {
+        self.observe_function_call("rdbms::mysql::db-connection", "begin-transaction");
+
+        let begin_oplog_index = self
+            .begin_durable_function(&DurableFunctionType::WriteRemoteBatched(None))
```

Comment:
`begin_durable_function` must be paired with an `end_durable_function` on _all_ code paths otherwise it is leaking. I think now it is only ended in case of some errors but not on the happy path. (I may be wrong, lots of lines to review)

Reply:
currently `end_durable_function` is invoked `HostDbTransaction.drop`, but will add all also handling to  `HostDbTransaction.commit` and `HostDbTransaction.rollback`